### DISCUSSION TITLE
What's the accepted/simplest way to create(cache) a Tween or sequence without starting it?

### DISCUSSION THREAD

**Discussion by ZedDevStuff** [KyryloKuzyk/PrimeTween#88] [FAQ] [answered]
I'm doing a kind of animation editor for my game and it needs to be able to not start itself when i create the Tween at the start or the end

**Comment by KyryloKuzyk** [comment, marked as answer]
In PrimeTween there is no way to create an animation without starting it. You can create an animation and pause it, but this is probably not what you're looking for.

If you're looking for DOTween's way of creating animation upfront, then calling PlayForward/Backwards() on it, there is no equivalent to it in PrimeTween for many reasons. Instead, in PrimeTween you create an animation at the same exact moment you need to play it.

More info:
https://github.com/KyryloKuzyk/PrimeTween?tab=readme-ov-file#tweenplayforwardplaybackwardsrestart
https://discussions.unity.com/t/primetween-high-performance-animations-and-sequences/926420/125

**Reply by ZedDevStuff** [reply]
If i do create a tween then pause it just after (the next line), will that have the time to do anything? Or would i have to reset the progress/time just to be sure?

**Reply by KyryloKuzyk** [reply]
Animations in PrimeTween receive their first update later in the frame (in LateUpdate to be exact), so yes, if you pause an animation right after you create it, it will stay in the initial state. No need to set progress to 0f manually in this case.

But I would recommend reconsidering whether you really want to do it because it will create an unnecessary complexity in your game. Why not create an animation at the exact moment you need to run it?  PrimeTween can start thousands of tweens and sequences with no performance overhead, so there is no need to create them and pause ahead of time.

**Comment by SimonNordon4** [comment]
Seeing as Tweens play when they are created, you can 'store' them inside a Func and create them when you need them.

```csharp
public class TestTweener : MonoBehaviour
{
    // We can 'store' the tween in a variable
    public Func<Tween> MyTween;

    public void Start()
    {
        // Assign the tween as an anonymous function.
        MyTween= () => Tween.PositionX(transform, 1f, new() { duration = 1f });    
    }

    public void PlayTween()
    {
         // Create a new tween when you're ready to use it.
         MyTween();
    }
}
```

I'm doing this because I want a central tween controller than I can add tweens to using composition. Although I'm assuming there some dangers about this. No chance to recycle tweens and I'm not sure about leaks.

```csharp
[RequireComponent(typeof(MeshRenderer))]
public class AnimationController : MonoBehaviour
{
    private MeshRenderer _meshRenderer;
    private TweenComponent[] _tweens;

    private IEnumerator Start()
    {
         _tweens = GetComponents<TweenComponent>();
         _meshRenderer = GetComponent<MeshRenderer>();

         var startSequence = Sequence.Create(cycles:3);
         foreach (var tween in _tweens)
         {
              startSequence.Group(tween.Create());
         }
         startSequence.OnComplete(() => _meshRenderer.enabled = false);              
     }
}
```

and my components

```csharp
public abstract class TweenComponent : MonoBehaviour
{
    public abstract Tween Create();
}

public class MoveInY : TweenComponent
{
    [SerializeField]
    private TweenSettings<float> MoveInYSettings = new()
    {
        startValue = -5f,
        endValue = 0f,
    };

    public override Tween Create() => Tween.PositionY(transform, MoveInYSettings);
}
```

Is this cursed? Time will tell.

**Reply by KyryloKuzyk** [reply]
> Seeing as Tweens play when they are created, you can 'store' them inside a Func and create them when you need them.

This approach doesn't cache the tween, it only caches the delegate that creates it. You can call `Tween.PositionX()` directly and the behavior would be the same:
```csharp
public class TestTweener : MonoBehaviour
{
    public void PlayTween()
    {
        // Create a new tween when you're ready to use it.
        Tween.PositionX(transform, 1f, new() { duration = 1f });
    }
}
```

> I want a central tween controller than I can add tweens to using composition. Although I'm assuming there some dangers about this. No chance to recycle tweens and I'm not sure about leaks.

The `TweenComponent` sample you provided is fine and doesn't introduce any dangers or memory leaks. You still don't achieve caching with this approach, but as I said before, you don't need to think about tween caching with PrimeTween. Just create animations when you need them.
